keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,674 @@
+
+### Non-Goals
+
+* Dynamically adjust system reserved and kube reserved values.
+* Hot unplug of node resources.
+* Update the autoscaler to utilize resource hot plugging.
+
+
+## Proposal
+
+This KEP aims to support the node resource hot plugging by adding a polling mechanism in kubelet to fetch the machine-information from cAdvisor's cache which is already updated periodically, This information will be fetched periodically by kubelet, after which the node status updater is responsible for updating this information at node level in the cluster.
+
+Additionally, this KEP aims to improve the initialization and reinitialization of resource managers, such as the memory manager and CPU manager, so that they can adapt to change in node's configurations.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I must be able to increase the cluster resource capacity without adding a new node to the cluster.
+
+#### Story 2
+
+As a cluster admin, I must be able to increase the cluster resource capacity without need to restarting the kubelet.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+1. Node resource hot plugging is an opt-in feature, merging the
+   feature related changes won't impact existing workloads. Moreover, the feature
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+2. Though the node resource is updated dynamically, the dynamic data is fetched from cAdvisor and its well integrated with kubelet.
+3. Resource manager are updated to adapt to the dynamic node reconfigurations, Enough tests should be added to make sure its not affecting the existing functionalities.
+
+## Design Details
+
+
+Below diagram is shows the interaction between kubelet and cAdvisor.
+
+
+```mermaid
+sequenceDiagram
+    participant node
+    participant kubelet
+    participant cAdvisor-cache
+    participant machine-info
+    kubelet->>cAdvisor-cache: Poll
+    cAdvisor-cache->>machine-info: fetch
+    machine-info->>cAdvisor-cache: update
+    cAdvisor-cache->>kubelet: update
+    kubelet->>node: node status update
+    kubelet->>node: re-initialize resource managers
+```
+
+The interaction sequence is as follows
+1. Kubelet will be polling in interval to fetch the machine resource information from cAdvisor's cache, Which is currently updated every 5 minutes.
+2. Kubelet's cache will be updated with the latest machine resource information.
+3. Node status updater will update the node's status with the latest resource information.
+4. Kubelet will reinitialize the resource managers to keep them up to date with dynamic resource changes.
+
+Note: In case of increase in cluster resources, the scheduler will automatically schedule any pending pods.
```

Comment:
Is the increase of resources trustworthy? What if kubelet got restarted and now it believes that the node size is smaller while scheduler already has a new higher value? Should we change kubelet initialization to "trust" the API server more than cAdvisor?

Reply:
shouldn't the kubelet attempt to update the node object in that case with the smaller value? ultimately we want the capacity to be right sized. if the scheduler thinks there's more memory than there is that won't help anyone

Reply:
we really need to write down on what are the expectations and behaviors today. I am not sure that the assumptions that node can change it's size and everything will keep working normally is true today.

Reply:
Big +1. The source of truth about machine data and its behavior (e.g. no downsizing events ever) is the cornerstone of this work and we need to be really careful here.

Reply:
What you described may work. But I am not sure what kind of side effects it may have. Simply stating that this is a supported scenario is a KEP on it's own - identify which tests to run and documentation to update.

Reply:
Sure, We have identified few test scenarios and added them under Test Plan.